### PR TITLE
Add Dynamic Battery Support for Base-Lock Profiled Door Lock Devices 

### PR DESCRIPTION
**Pull request by hcarter-775**
Check all that apply

# Type of Change

- [ ] WWST Certification Request
     - If this is your first time contributing code:
          - [ ] I have reviewed the README.md file
          - [ ] I have reviewed the CODE_OF_CONDUCT.md file
          - [ ] I have signed the CLA
     - [ ] I plan on entering a WWST Certification Request or have entered a request through the WWST Certification console at developer.smartthings.com
- [ ] Bug fix
- [x] New feature
- [ ] Refactor

# Checklist

- [x] I have performed a self-review of my code
- [x] I have commented my code in hard-to-understand areas
- [ ] I have verified my changes by testing with a device or have communicated a plan for testing
- [x] I am adding new behavior, such as adding a sub-driver, and have added and run new unit tests to cover the new behavior

# Description of Change
Jira ticket showing mis-profiling of a bridged door lock: https://smartthings.atlassian.net/browse/MTR-821
Further, there is a general ticket on adding dynamic profiling to this driver: https://smartthings.atlassian.net/browse/CHAD-13948

This logic will permit dynamic profiling on the base-lock profile depending on the Power Source cluster, and will ignore devices that do not auto-profile as base-lock.

# Summary of Completed Tests
Unit tests
Bridged Switchbot device - Passed
VDA Door Lock - Passed
VDA Door Lock (Lock Codes) - Passed
Yale Assure Matter Lock - Passed

### CODE REVIEW

File: drivers/SmartThings/matter-lock/src/init.lua
Comment (nickolas-deboom) [inline drivers/SmartThings/matter-lock/src/init.lua:520]: A new fingerprint was added yesterday and should be added to this list: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/1650

**Comment by nickolas-deboom** [comment]
This looks good to me, but I was just thinking that one potential drawback with this approach is that each PR for new WWST device fingerprints would also have to include an update to the list of fingerprinted devices in `do_configure`. Maybe that is no big deal but I thought I'd mention it

**Comment by hcarter-775** [comment]
Yeah @nickolas-deboom I agree, that's definitely a downside. Unfortunately, there's not really any better method in place of solving this issue atm. I figured we would inform Alissa of this in the next Risk Analysis meeting and then you and I can help monitor this as well until we put a better solution in place.

**Comment by hcarter-775** [comment]
@nickolas-deboom I also think this example might get some more proactive desire to fix this situation in a larger way

File: drivers/SmartThings/matter-lock/src/init.lua
Comment (ctowns) [reply drivers/SmartThings/matter-lock/src/init.lua:520]: I think Nick makes a very good point about this requiring an update to this list every time we add a WWST fingerprint. I don't think this would be very difficult, but we have received push back on this in the past due to the way it complicates the WWST process. I think if we communicate with @lelandblue and make sure this is coordinated with the WWST team, this could be an extra step that is added to the WWST process and we can maintain this list here ourselves.

However, I think there is a different way to think about issue as well. Currently, we are trying to filter out all devices that have a fingerprint because these already have assigned profiles that we do not want to change. However, the assumption we are making here is that all non-fingerprinted devices are using the `base-lock` profile, which is true in this case.

So, what if instead of filtering out fingerprinted devices, we instead just specifically targeted devices using the `base-lock` profile? This would be a different way to get the same subset of devices, and it would prevent the need to maintain a list of fingerprinted devices in the driver. 

One way we can do this is by using the `device:supports_capability()` function to determine if a device supports a given capability, and we can call this function multiple times for each of the capabilities in the `base-lock` profile to see if the device supports all the capabilities in the `base-lock` profile. If it does, we can assume that it is using the `base-lock` profile, because based on my initial review of the profiles, the `base-lock` profile does not contain capabilities that are a subset of any other profile, so therefore if a device supports all the capabilities in `base-lock`, it is definitely using the `base-lock` profile. I have done something similar to this for the `matter-sensor` driver here, where we essentially reconstructed a profile based on the capabilities that the device supports and then we append `-battery` on the end if the device also supports battery: 
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/967/files#diff-8a122f65a4419a8fe00de96e7444d9d71df5609075dbb217531dc6d125197755R26-R36

I think we could do something similar here where we do something like this:
```
  if device:supports_capability(capabilities.lock) and
    device:supports_capability(capabilities.lockCodes) and
    device:supports_capability(capabilities.tamperAlert) and
    device:supports_capability(capabilities.battery) then
    -- we can assume device is using base-lock profile
    -- handle switching to no battery profile as needed
  end
```

Now, we know there will be changes coming to the `matter-lock` driver that will make this assumption of "non-fingerprinted devices use `base-lock`" a little more complicated, but I think we could use the same method of essentially constructing the profile from it's capabilities that is used above to handle things more granularly once we need to consider the profiles that support `lockAlarm` with the new matter lock driver.

Comment (hcarter-775) [reply drivers/SmartThings/matter-lock/src/init.lua:520]: Good catch!

**Comment by hcarter-775** [comment]
And I just thought of this, but another point of note here is that there is a new and improved matter Lock driver coming soon, so this issue may be extremely temporary.

**Comment by nickolas-deboom** [comment]
Those are good points, and I can't think of a better way to handle this either, so I think this is a good solution for now, especially given it might be temporary since there are improvements coming to this driver as you mentioned.

**Review by nickolas-deboom — APPROVED** [review]

File: drivers/SmartThings/matter-lock/src/init.lua
Comment (hcarter-775) [reply drivers/SmartThings/matter-lock/src/init.lua:520]: Yeah @ctowns I think this seems like a very reasonable idea that does the same thing without needing an extra table!

**Review by ctowns — COMMENTED** [review]
This looks good, however, there are multiple other places in the `matter-lock` driver that update the device's profile, for example:
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/44087fd33c74c98dcfae8acd5e2dc5b7cc498aaa/drivers/SmartThings/matter-lock/src/init.lua#L573
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/44087fd33c74c98dcfae8acd5e2dc5b7cc498aaa/drivers/SmartThings/matter-lock/src/init.lua#L200
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/44087fd33c74c98dcfae8acd5e2dc5b7cc498aaa/drivers/SmartThings/matter-lock/src/init.lua#L213

Two of the instances have to do with setting a profile for a non-functional lock, however in the `device_added` function, I think we'd also need to do something similar in the case that the device is going to be switched to the `lock-without-codes` profile. Essentially, you'd have to just add a check that the device would switch to `lock-without-codes-batteryLevel/nobattery` depending on what it supports like you have done above.

Also, when recovering from a non-functional state, we'd probably want to copy what you've done and make sure the device joins to `base-lock` plus whatever is appropriate for it's battery situation.

**Comment by ctowns** [comment]
Also before merging, this change should be tested with both the new bridged the device as well as regression tested against an existing lock.

**Comment by hcarter-775** [comment]
Yes, I agreee Cooper. I will add the dynamic battery profiling logic to those places as well. And I have already reached out on the affected ticket to get some testing with the new device.

**Review by ctowns — APPROVED** [review]
Looks good! I would suggest waiting for the new matter lock PR to merge first and then rebasing these changes on top of it. This PR should be easier to rebase and we want the new matter lock PR to be as frictionless as possible a this point.

**Review by Huangxiangjie — COMMENTED** [review]
If device Power Source 0x002F Features is BAT, But do not support 0x000D BatTimeRemaining

**Pull request merged**
